keps/sig-api-machinery/NNNN-watch-list/README.md

```diff
@@ -0,0 +1,956 @@
+After that, the interval will be put back to the previous value.
+
+Note: Since we get a notification every 5 seconds from etcd and we try to deliver a bookmark every 1 second.
+It seems the maximum delay time a reflector will have to wait after receiving initial data is 6 seconds (assuming small dataset).
+It might be unlikely in practice since we might get bookmarkAfterResourceVersion even before handling initial data.
+Also sending data itself takes some time as well.
+
+Step 3: After receiving a BOOKMARK event the reflector is considered to be synchronized.
+It replaces its internal store with the collected items (syncWith) and reuses the current connection for getting further events.
+
+### Important optimisations
+
+1. Avoid DeepCopying of initial data<br><br>
+The watchCache has an important optimization of wrapping objects into a cachingObject.
+Given that objects aren't usually modified (since selfLink has been disabled) and the fact that there might be multiple watchers interested in receiving an event.
+Wrapping allows us for serializing an object only once.
+![removing deep copying from the watch cache](./cacher_deep_copy.png "removing deep copying of object from the watch cache")
+The watchCache maintains two internal data structures. 
+The first one is called the store and is driven by the reflector.
+It essentially mirrors the content stored in etcd.
+It is used to serve LIST requests.
+The second one is called the cache, which represents a sliding window of recent events received from the reflector.
+It is effectively used to serve WATCH requests from a given RV.<br><br>
+By design cachingObjects are stored only in the cache.
+As described in Step 2, the cacheWatcher gets initial data from the watchCacher.
+The latter, in turn, gets data straight from the store.
+That means initial data is not wrapped into cachingObject and hence not subject to this existing optimization.<br><br>
+Before sending objects any further the cacheWatcher does a DeepCopy of every object that has not been wrapped into the cachingObject.
+Making a copy of every object is both CPU and memory intensive. It is a serious issue that needs to be addressed.<br><br>
+With RemoveSelfLink [graduating](https://github.com/kubernetes/kubernetes/blob/956dfb5196c2884b61804e94810e29ca077301ee/staging/src/k8s.io/apiserver/pkg/features/kube_features.go#L138) to GA (and already disabled by default) we are able to safely avoid this copying.
+2. Reduce the number of allocations in the WatchServer<br><br>
+The WatchServer is largely responsible for streaming data received from the storage layer (in our case from the cacher) back to clients.
+It turns out that sending a single event per consumer requires 4 memory allocations, visualized in the following image.
+Two of which deserve special attention, namely the allocations 1 and 3 because they won't reuse memory and rely on the GC for cleanup.
```

Comment:
Did you try running your experiment with JSON as an output format (as opposed to protobuf)?

JSON is obviously more verbose, but I'm wondering if it is doing that many allocations too (from quick glance into the code it seems to behave better).

Reply:
nope, but I optimized this part and run the test with `2048` informers - all synced! (1 TB of data). Here is a graph of memory usage during that run:

![2048-informers](https://user-images.githubusercontent.com/21304151/149896016-8bfb188c-9888-463f-932e-20eea87947eb.png)